File: apps/api/schemas/service.py
"""Pydantic v2 schemas for service catalog, baseline profiles, and graph topology."""

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ServiceResponse(BaseModel):
    """Registered microservice profile schema."""

    model_config = ConfigDict(from_attributes=True)

    name: str
    service_type: str
    capacity: int
    baseline_latency_ms: float
    baseline_failure_rate: float
    timeout_ms: float
    max_retries: int
    retry_backoff_ms: float
    dependencies: list[Any] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class ServiceUpdate(BaseModel):
    """Payload for modifying a service's baseline parameters."""

    capacity: int | None = Field(default=None, ge=1, le=10000)
    baseline_latency_ms: float | None = Field(default=None, ge=0.0)
    baseline_failure_rate: float | None = Field(default=None, ge=0.0, le=1.0)
    timeout_ms: float | None = Field(default=None, ge=1.0)
    max_retries: int | None = Field(default=None, ge=0, le=10)
    retry_backoff_ms: float | None = Field(default=None, ge=0.0)
    dependencies: list[Any] | None = Field(default=None)
    metadata: dict[str, Any] | None = Field(default=None)


class ServiceLatencyStatsResponse(BaseModel):
    """Database-side latency percentile distribution response schema."""

    service: str
    count: int
    mean_latency_ms: float
    median_p50_latency_ms: float
    p90_latency_ms: float
    p95_latency_ms: float
    p99_latency_ms: float
    min_latency_ms: float
    max_latency_ms: float


class ServiceHealthResponse(BaseModel):
    """Service operational reliability and error rate summary schema."""

    service: str
    total_events: int
    failure_count: int
    error_rate_percent: float
    retry_count: int
    retry_rate_percent: float
    timeout_count: int
    timeout_rate_percent: float
    latency: ServiceLatencyStatsResponse


class TopologyNode(BaseModel):
    """Graph topology node representing a service or infrastructure component."""

    id: str = Field(description="Service identifier")
    name: str = Field(description="Display name")
    type: str = Field(description="Service category (BUSINESS, CACHE, DATABASE, GATEWAY)")
    capacity: int = Field(description="Concurrency worker capacity")
    baseline_latency_ms: float = Field(description="Baseline response latency in ms")


class TopologyEdge(BaseModel):
    """Directed dependency edge between two services."""

    from_service: str = Field(description="Caller / upstream service")
    to_service: str = Field(description="Callee / downstream dependency")
    relationship_type: str = Field(
        description="Dependency type (HTTP_RPC, CACHE_LOOKUP, DB_QUERY, GATEWAY_CALL)"
    )
    call_weight: float = Field(default=1.0, description="Relative call frequency or weight")
    metadata: dict[str, Any] = Field(default_factory=dict)


class ServiceTopologyResponse(BaseModel):
    """Complete service graph topology response for graph visualizers."""

    nodes: list[TopologyNode]
    edges: list[TopologyEdge]
    total_services: int
    total_dependencies: int

File: frontend/src/types/service.ts
/**
 * Service domain and telemetry types matching backend Pydantic schemas.
 */

export interface ServiceProfile {
  name: string;
  service_type: string;
  capacity: number;
  baseline_latency_ms: number;
  baseline_failure_rate: number;
  timeout_ms: number;
  max_retries: number;
  retry_backoff_ms: number;
  dependencies: (string | { to: string; type?: string; weight?: number })[];
  metadata: Record<string, unknown>;
}

export interface ServiceUpdatePayload {
  capacity?: number;
  baseline_latency_ms?: number;
  baseline_failure_rate?: number;
  timeout_ms?: number;
  max_retries?: number;
  retry_backoff_ms?: number;
}

export interface ServiceLatencyStats {
  service: string;
  count: number;
  min_latency_ms: number;
  max_latency_ms: number;
  mean_latency_ms: number;
  median_p50_latency_ms: number;
  p90_latency_ms: number;
  p95_latency_ms: number;
  p99_latency_ms: number;
  start_time?: string | null;
  end_time?: string | null;
}

export interface ServiceHealth {
  service: string;
  total_events: number;
  successful_events: number;
  failed_events: number;
  timeout_events: number;
  retry_events: number;
  failure_rate_percent: number;
  timeout_rate_percent: number;
  retry_rate_percent: number;
  avg_latency_ms: number;
  start_time?: string | null;
  end_time?: string | null;
}

export interface TopologyNode {
  id: string;
  name: string;
  type: string;
  capacity: number;
  baseline_latency_ms: number;
}

export interface TopologyEdge {
  from_service: string;
  to_service: string;
  relationship_type: string;
  call_weight: number;
  metadata: Record<string, unknown>;
}

export interface ServiceTopology {
  nodes: TopologyNode[];
  edges: TopologyEdge[];
  total_services: number;
  total_dependencies: number;
}

export interface ServiceHealthSummary {
  service: string;
  total_events: number;
  error_count: number;
  error_rate_percent: number;
  retry_count: number;
  retry_rate_percent: number;
  timeout_count: number;
  timeout_rate_percent: number;
  mean_latency_ms: number;
  p95_latency_ms: number;
}
